### ISSUE TITLE
Smarty cache issue with getDevice() and isMobile() in module template

### ISSUE THREAD

**Issue by szykuc** [PrestaShop/PrestaShop#21661]
#### Describe the bug
Smarty conditions Context::getContext()->isMobile() and getDevice() don't work when smarty cache is enabled

#### Expected behavior
In module template i use condition {if Context::getContext()->isMobile() == 1} and i want show different text for mobile and dekstop.
When smaty cache is disabled everything works perfect.

#### Steps to Reproduce
Set smarty cache to disabled

Make condition in template e.g. 
/themes/modules/ps_newproducts/views/templates/hook/ps_newproducts.tpl
or direct in module if no file in theme
/modules/ps_newproducts/views/templates/hook/ps_newproducts.tpl

{if Context::getContext()->isMobile() == 1}
    mobile
{else}
    desktop
{/if}
Check if working. Should display 'mobile' on mobile device and 'desktop' on dekstop.
Set smarty cache to on, clear it and refresh page on both devices.

#### Additional information
PrestaShop version: 1.7.6.8
PHP version: 7.1.33 and 7.2

**Comment by hibatallahAouadni** [comment]
hi @szykuc 

I managed to reproduce your issue with PS1.7.5.2, PS1.7.6.8 and PS1.7.7.0-rc.1+build.3, even when i set "Template Compilation" to "Force compilation"
![image](https://user-images.githubusercontent.com/23308427/97677470-973f7d80-1a92-11eb-91d6-0a494c6899e4.png)

This problem doesn't occur when you make this condition in template e.g. /themes/classic/templates/index.php
It only happen when you put this condition under any native module like:  /themes/modules/ps_featuredproducts/views/templates/hook/ps_featuredproducts.tpl

I’ll add this to the debug backlog so that it’s fixed.
If you have already fixed it on your end or if you think you can do it, please do send us a pull request!

Thanks!

**Comment by sellmark** [comment]
I was able to reproduce this in core template file: /themes/classic/templates/index.tpl so it seems like a global issue, not only modules-related.

**Comment by hugprat** [comment]
Hi, we have the same problem here, Prestashop 1.7.6.2. can't enable Smarty Cache because we use the isMobile function in various places. Any help will be welcome.

Thanks.

**Comment by infinity1234** [comment]
Hi, same problem here, Prestashop 1.7.8.1

**Comment by vallka** [comment]
Same problem, noticed from PS 1.7...0, currently 1.7.8.1. Cannot use getDevice() and isMobile() etc. functions in modules(only!) templates. I seem to find a solution, will try and post it here if it is working....

**Comment by vallka** [comment]
Ok, my fix is working. To fix it permanently, 1 line of code is needed in /classes/module/Module.php in method getCacheId (line 2406):
```
    protected function getCacheId($name = null)
    {
        $cache_array = [];
        $cache_array[] = $name !== null ? $name : $this->name;
        if (Configuration::get('PS_SSL_ENABLED')) {
            $cache_array[] = (int) Tools::usingSecureMode();
        }
        if (Shop::isFeatureActive()) {
            $cache_array[] = (int) $this->context->shop->id;
        }
        if (Group::isFeatureActive() && isset($this->context->customer)) {
            $cache_array[] = (int) Group::getCurrent()->id;
            $cache_array[] = implode('_', Customer::getGroupsStatic($this->context->customer->id));
        }
        if (Language::isMultiLanguageActivated()) {
            $cache_array[] = (int) $this->context->language->id;
        }
        if (Currency::isMultiCurrencyActivated()) {
            $cache_array[] = (int) $this->context->currency->id;
        }
        $cache_array[] = (int) $this->context->country->id;
        
        // added by vallka:
        $cache_array[] = (int) $this->context->getDevice();

        return implode('|', $cache_array);
    }

```

Prestashop 1.7.8.1

**Comment by vallka** [comment]
I never tried (yet) to add my updates to main Prestashop source, all these 'pull requests', how to do it properly... Meanwhile I can fix it for myself and others who need it now, by writing a small override module:
/override/classes/module/Module.php :

```
<?php

class Module extends ModuleCore
{
    protected function getCacheId($name = null)
    {
        $result = parent::getCacheId($name);

        return $result . '|' .  Context::getContext()->getDevice();
    }
}

```

**Comment by eternoendless** [comment]
Adding the device to the cacheId would triple the amount of cache files (if you get visits from all three types of devices) and would only be useful if you wanted to serve different content to a specific device type. I think that nowadays there are better options with CSS media queries, etc.

Modules still have full control over the way their cacheId is made up, so to me this looks like a "won't fix", unless I'm missing something.

**Comment by lim3ra** [comment]
This problem is still present in Presta 8.1.5

anyone has successfully tested this solution on prod? Is there any negative impact caused by larger amount of cache files?

**Comment by Blaaszkovsky** [comment]
This error also occurs in version 8.0.5 when using ex. `Context::getContext()->isMobile()`. The solution is to add a separate cache module for mobile and a separate one for desktop. For example on ps_customeraccountlinks:
```php
if (!$this->isCached($this->templateFile, $this->getCacheId($tpl_key . '_' . ($this->context->isMobile() ? 'mobile' : 'desktop')))) {
    $this->smarty->assign($this->getWidgetVariables($hookName, $configuration));
}
```
```php
if (!$this->isCached($this->templateFile, $this->getCacheId($tpl_key . '_' . ($this->context->isMobile() ? 'mobile' : 'desktop')))) {
    $this->smarty->assign($this->getWidgetVariables($hookName, $configuration));
}
```

Then a separate cache is generated for the mobile and desktop views, and the problem disappears.

**Comment by boo-code** [comment]
Still reproducible on 9.2.x, and the one-line diff from #27068 is the right place.

`Module::getCacheId()` builds the key from what a rendered block can depend on — secure mode, shop, customer group, language, currency, country. The device is not among them, so a template branching on `isMobile()` or `getDevice()` gets whichever variant was cached first served to every visitor.

The fair objection to adding it is cost, so here is where I landed. Every other entry is gated on that dimension being able to vary, `Shop::isFeatureActive()`, `Language::isMultiLanguageActivated()` and so on. There is no equivalent switch for the device: `PS_ALLOW_MOBILE_DEVICE` no longer exists in 9.x, and `Context::getDevice()` always resolves to computer, tablet or mobile from the user agent, so the analogous condition is always true.

That means up to three cached entries per module block instead of one, on shops that enable the Smarty cache. On a responsive theme that does not branch those entries are identical, so the cost is storage. On a theme that does branch, the single entry today is wrong for two visitors out of three.

Written that way, with a test pinning that the three devices produce three ids, that one device produces a stable id, and that the id still varies by block name. If you would rather pay the cost only where it is needed, the alternative is a configuration switch read here the way the others are — the same line behind a condition — and I will write it that way instead if you prefer.

Credited to @vallka, whose #27068 had the same line.

I am at my open pull request limit, so I will link it as soon as I have a slot.
